Export dialog in DrILL

gui-co opened the pull request:
**Description of work.**

This PR adds an automatic export system to the DrILL interface. When the reduction of a sample is done, the automatic export is triggered and exported files are written to the default save directory.

Each acquisition mode has a number of export algorithm associated to it. Some of them are activated by default, some are not. The activated algorithms can be seen and changed through a dedicated dialog in DrILL interface.

**To test:**

Export dialog:
* open the DrILL interface
* open the export dialog (mdi-application-export icon in the tool bar - https://pictogrammers.github.io/@mdi/font/3.6.95/)
* observe the selected export algorithm

Export:
With data in `SystemTest/ILL/D11/`:
* check your default save directory
* open the DrILL interface, select D11 instrument
* put 2888 in the column `SampleRuns` of the first row
* process that row
* check in your default save directory that the automatic export produced the right files
* change the selected algorithm in the export dialog, process again and observe the files in your save directory

Part of #29570

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson commented:
@gui-co - was just looking through the "ready for review" PR list and found this one. I can review if you like. Please can you add some testing instructions and fix the merge conflict?

gui-co commented:
@DannyHindson, thanks for the review. I resolved the conflicts and added a description + tests in the PR.

DannyHindson commented on scripts/Interface/ui/drill/model/DrillExportModel.py:
```diff
@@ -0,0 +1,103 @@
+                    if s:
+                        filename = exportPath + wsName \
+                                   + RundexSettings.EXPORT_ALGO_EXTENSION[a]
+                        task = DrillTask(-1, a, InputWorkspace=wsName,
```
Does the ref need to be something other than -1 here so that multiple export tasks can be distinguished from each other later on?

gui-co replied:
The DrillTask ref is used to find out from which task success or error signals are received. This is used during the reduction to report the status on the table. But as I said in the comment, here we are ignoring these signals.

DannyHindson reviewed (changes requested):
Thanks for the extra details. I've just tested this and I've got one problem. The SaveNISTDAT algorithm doesn't seem to be producing a file. The files for the other SaveXXX algorithms work OK.

I've debugged it a bit and I think this is happening because the SaveNISTDAT algorithm is producing this error:

_Invalid value for property InputWorkspace (MatrixWorkspace) from string "sample_1_1": The workspace must contain histogram data_

The workspace looks like it has point data rather than histogram data. This error doesn't get recorded in the workbench messages log though, I just see this in the debugger. So it looks like the failure situation isn't being managed properly. Not sure what code handles the taskError signal

gui-co commented:
The export algorithms are acquisition mode specific but depending on some reduction parameters some of them can not be applied. Because the interface should stay agnostic of the reduction, we decided to run all the algorithms and silently ignore the export errors.

I think this is fine as long as users know that some exports are valid with only certain types of data (e.g. SaveNISTDAT that needs I(Qx, Qy) output type and not I(Q), the default option).

DannyHindson commented:
> The export algorithms are acquisition mode specific but depending on some reduction parameters some of them can not be applied. Because the interface should stay agnostic of the reduction, we decided to run all the algorithms and silently ignore the export errors.
> 
> I think this is fine as long as users know that some exports are valid with only certain types of data (e.g. SaveNISTDAT that needs I(Qx, Qy) output type and not I(Q), the default option).

OK I see. I think with that approach you will also hide any errors which aren't caused by the export algorithm being inapplicable eg if the user doesn't have permission on their save directory. Would it be better to log a debug message saying "export ignored" with the reason?
I can see that this is a design choice for an ILL-specific interface so if you feel strongly that it should be left as it is then let me know and we can go with it as is.

gvardany commented:
We should definitely log an information if a save act fails. What we could also do, is in the dialog add the matching output types next to each algorithm, let's say `SaveCansas1D (I(Q))`, so that users don't get surprised if it is not produced when doing I(Qx,Qy).

gvardany commented:
However the error that you got when debugging NISTDAT is only about histogram/point data, but NISTDAT is fundamentally for I(Qx,Qy) only.

gui-co commented:
Runtime errors logged by the algorithm are not silenced. It means that the write access error that you mentioned will be visible to the user (I checked that for SaveCanSAS1D).

In the export dialog, I added a tool button for each algorithm that is linked to its documentation. This will help users decide which export are compatible with their processing. I also updated the DrILL documentation and added a release note.

DannyHindson commented:
> Runtime errors logged by the algorithms are not silenced, only the exceptions thrown by `setProperty()` and `execute()` methods. It means that the write access error that you mentioned will be visible to the user (I checked that for SaveCanSAS1D).
> 
> In the export dialog, I added a tool button for each algorithm that is linked to its documentation. This will help users decide which export are compatible with their processing. I also updated the DrILL documentation and added a release note.

Re the silencing of the errors - I tried running the automatic export with an invalid "default save directory" and I don't get an error message. Please can you try this?

Also if some of the error signals are being processed wouldn't you need to change the ref to something other than -1 so you can see which task the error came from?

gui-co commented:
The problem is that the errors are thrown and/or logged at different level:
* if the save directory does not exist: it throw during `setProperty()`
* if the directory exists but is read-only: it logs an error and throws during `execute()`
* if the input workspace is not adapted: it throws during `setProperty()`

I will see what needs to be logged and come back with a commit.

DannyHindson reviewed (changes requested):
I've just seen you've added a couple more commits to this thanks.
I can see you've added an extra validation in the DrillExportModel relating to the save directory. It feels like it should be possible to expose any exceptions raised during the execution of the SaveXXX algorithms rather than having to replicate the SaveXXX validation logic in the DrillExportModel.
At the moment there is nothing in the DrillExportModel that listens for the task complete\task error signals from the algorithms in the DrillAlgorithmPool. If you did something like the following in DrillExportModel.py you would capture any errors as an information message including the "invallid directory" error. This is based on the code in DrillModel:
```
    def __init__(self, acquisitionMode):
        """
        Create the export model by providing an aquisition mode.

        Args:
            acquisitionMode (str): acquisition mode
        """
        self._exportAlgorithms = {k:v
                for k,v
                in RundexSettings.EXPORT_ALGORITHMS[acquisitionMode].items()}
        self._pool = DrillAlgorithmPool()

        self._pool.signals.taskError.connect(self._onTaskError)

    def _onTaskError(self, ref, msg):
        """
        Called when a processing fails. This method logs a message

        Args:
            ref (int): sample index
            msg (str): error msg
        """
        name = str(ref + 1)
        logger.information("Error while processing sample {0}: {1}"
                     .format(name, msg))
```
At the moment this will only work if you run a single export because the task refs are all -1 (so the signals don't distinguish between tasks). I tried it and it does show an information message "directory doesn't exist" in the log if you change the save directory to a nonsense value

gui-co commented:
> It feels like it should be possible to expose any exceptions raised during the execution of the SaveXXX algorithms rather than having to replicate the SaveXXX validation logic in the DrillExportModel.
At the moment there is nothing in the DrillExportModel that listens for the task complete\task error signals from the algorithms in the DrillAlgorithmPool. If you did something like the following in DrillExportModel.py you would capture any errors as an information message including the "invallid directory" error.

This is what I wanted not to do, to avoid logging of errors related to an incompatible algorithm.

So I added a way to define a boolean expression in the configuration that will activate the corresponding export algorithm based on the value of processing parameters. With this, only the compatible algorithms will be executed on the data.

All the signals are now connected, the user will get all the errors (error channel) and a summary of the successes at the end (notice channel).

DannyHindson commented on docs/source/interfaces/DrILL.rst:
```diff
@@ -194,6 +194,25 @@ row(s). To get further information about the errors, one has to look into the
+.. image:: ../images/drill-export.png
+   :align: center
+
+For each acquision mode, a list of adapated algorithms will be displayed in that
```
I think this should be "acquisition mode"

gui-co replied:
Done.

DannyHindson commented on docs/source/interfaces/DrILL.rst:
```diff
@@ -194,6 +194,25 @@ row(s). To get further information about the errors, one has to look into the
+dialog. Some of them are activated by default but the user is free to select the
+ones he wants. All checked algorithms will be applied on all output workspaces
+of all processed rows. The exported files are saved in the Mantid default save
+directory. If the algorithm is not adapted to the data, it will be skiped. For
```
skipped rather than skiped

gui-co replied:
Done.

DannyHindson commented on scripts/Interface/ui/drill/model/DrillExportModel.py:
```diff
@@ -0,0 +1,206 @@
+            if ((workspaceName in wsName)
+                    and (not isinstance(mtd[wsName], WorkspaceGroup))):
+                for a,s in self._exportAlgorithms.items():
+                    if s and self._validCriteria(wsName, a):
```
I think it would be nice to log a debug message saying "export skipped" or similar if the algorithm doesn't meet the criteria

gui-co replied:
I added a log in the notice channel for this.

DannyHindson reviewed (changes requested):
Thanks for your patience in making the changes to this. I've retested and the new scheme looks good and provides a nice balance between exposing real errors and not distracting the user with pointless errors about exports that aren't applicable. 
I've highlighted a couple of typos and one suggestion about logging a debug message if a selected export algorithm is skipped but hopefully complete after that.

gui-co commented:
Thanks for your patience. I corrected the typos in the doc and added a log message for skipped algorithms.
I used the notice channel for this log, the same as for the successful exports.

DannyHindson reviewed (approved):
Looks good now thanks

gvardany commented:
The only bit I'm not happy with right now is checking looping over all workspaces in ADS, but touching the ADS in the last minute is not a good idea. So could you open an issue to add a method to ADS to return workspaces matching the given name?
